### PR TITLE
Refactoring Dataset and transformations for multiple inputs

### PR DESCRIPTION
**Pull request by AnBucquet**
As mentioned in #4, the `__getitem__` method in the current dataset returns a single tensor, even if there are multiple modalities. This PR aims to return a list of tensors instead and adapt all transformations to that. 

**N.B.** The dataset returns a list of tensor even if there only one modality (ie a list of one element). But in that case, the transformation `ToTensor()` removes the list and return only the tensor.

### CODE REVIEW

File: medicaltorch/transforms.py [hunk 1]
```diff
@@ -111,6 +113,19 @@ def __call__(self, sample):
         return sample
 
 
+class StackTensors(MTTransform):
```
Comment (charleygros) [inline medicaltorch/transforms.py:118]: cool!!

File: medicaltorch/transforms.py [hunk 2]
```diff
@@ -140,13 +155,13 @@ def __init__(self, size, labeled=True):
...
     def propagate_params(sample, params):
-        input_metadata = sample['input_metadata']
+        input_metadata = sample['input_metadata'][0]
         input_metadata["__centercrop"] = params
         return input_metadata
 
     @staticmethod
     def get_params(sample):
-        input_metadata = sample['input_metadata']
+        input_metadata = sample['input_metadata'][0]
```
Comment (charleygros) [inline medicaltorch/transforms.py]: Index[0] --> first channel? Would it be a problem if the params (ie cropping coords) are not the same across channel?

Actually, not sure this case could happen. At least, will not happen with both `CenterCrop2D` and `ROICrop2D` since we currently have only one GT for all channels.

File: medicaltorch/transforms.py [hunk 3]
```diff
@@ -432,13 +451,13 @@ def __call__(self, sample):
         input_data = sample['input']
         gt_data = sample['gt'] if self.labeled else None
         if np.random.randint(2) == 1:
-            input_data = np.flip(input_data, axis=0).copy()
+            input_data = np.flip(input_data, axis=1).copy()
```
Comment (charleygros) [inline medicaltorch/transforms.py:495]: axis=1 since axis=0 represents the channels, okay, all good.

File: medicaltorch/datasets.py [hunk 4]
```diff
@@ -325,43 +331,53 @@ def __getitem__(self, index):
...
+            'input': input_tensors,
+            'gt': gt_img,
+            'roi': roi_img,
+            'input_metadata': input_metadata,
+            'gt_metadata': seg_pair_slice['gt_metadata'],
+            'roi_metadata': roi_pair_slice['gt_metadata']
+        }
 
+        """"
+        Moving that part in ToTensor() transformation
```
Comment (charleygros) [inline medicaltorch/datasets.py:374]: Good documentation thanks

**Review by charleygros — APPROVED** [review]
There was actually more changes needed that what I thought!! Thanks Anthime!

File: medicaltorch/transforms.py (hunk 2 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py]: I've made the assumption that all input_metadata are the same. Maybe I should change that?

File: medicaltorch/transforms.py (hunk 3 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py:495]: Yup! Looking if I can do that more often instead of using `for`  loop.

File: medicaltorch/datasets.py [hunk 5]
```diff
@@ -269,14 +274,16 @@ def _load_filenames(self):
...
-        """This method will replace the current transformation for the
+        """ This method will replace the current transformation for the
         dataset.
 
         :param transform: the new transformation
         """
         self.transform = transform
 
     def compute_mean_std(self, verbose=False):
+        # TODO: adapt to multi channels
```
Comment (charleygros) [inline medicaltorch/datasets.py]: At this point we are always in mono-channel, no? Am I missing something?

Comment (AnBucquet) [reply medicaltorch/datasets.py]: Oups you're right, forgot to change that! Mostly because this method is never called...  
Should we still adapt it ? if yes should we compute the mean/ std per channel or just for the entire dataset ??

File: medicaltorch/transforms.py (hunk 2 above)
Comment (charleygros) [reply medicaltorch/transforms.py]: I am just thinking: if the 2 contrasts do not have the same initial size (case scenario of Andréanne), then the cropping info will not be the same across channels.
Slice index, filename...etc. would also be different I think --> to double check.
If that 's the case, we should be carefull with that :/

Comment (AnBucquet) [reply medicaltorch/transforms.py]: Good point effectively... need to check that!

File: medicaltorch/datasets.py (hunk 5 above)
Comment (charleygros) [reply medicaltorch/datasets.py]: We should maybe use it... it could help (normalize with mean and std of the contrast-dataset).
Per channel makes more sense I think

Comment (AnBucquet) [reply medicaltorch/datasets.py]: Fixed in [bc1dbbb](https://github.com/neuropoly/medicaltorch/pull/9/commits/bc1dbbbdbf2f61f0795496ee2e13266820848b00).

Comment (charleygros) [reply medicaltorch/datasets.py]: I have seen it --> really really nice!

File: medicaltorch/datasets.py [hunk 6]
```diff
@@ -270,47 +277,54 @@ def _load_filenames(self):
...
                                  refresh=False)
 
-        training_std = np.sqrt(sum_var / numel)
-        return training_mean.item(), training_std.item()
+            training_std = np.sqrt(sum_var / numel)
+        # Converting tensors to numpy array
+        training_mean = [training_mean[i].item() for i in range(self.n_contrasts)]
+        training_std = [training_std[i].item() for i in range(self.n_contrasts)]
+        print("Training means are equal to {}".format(training_mean))
+        print("Training means are equal to {}".format(training_mean))
```
Comment (charleygros) [inline medicaltorch/datasets.py]: Duplication? also L307

File: medicaltorch/transforms.py [hunk 7]
```diff
@@ -46,21 +48,25 @@ def __call__(self, sample):
...
-        if isinstance(input_data, list):
-            ret_input = [F.to_tensor(item)
-                         for item in input_data]
+        if len(input_data) > 1:
+            # Multiple inputs
+            ret_input = [F.to_tensor(item) for item in input_data]
         else:
-            ret_input = F.to_tensor(input_data)
+            # single input
+            ret_input = F.to_tensor(input_data[0])
```
Comment (charleygros) [inline medicaltorch/transforms.py:56]: ret_input is either a list either a tensor --> does this cause no issue further in the code?

File: medicaltorch/transforms.py [hunk 8]
```diff
@@ -46,21 +48,25 @@ def __call__(self, sample):
...
+            # transform list of dic into single dic
+            rdict['input_metadata'] = sample['input_metadata'][0]
```
Comment (charleygros) [inline medicaltorch/transforms.py:58]: What about input_metadata for len>1?

File: medicaltorch/transforms.py (hunk 2 above)
Comment (charleygros) [reply medicaltorch/transforms.py]: Any updates here?

File: medicaltorch/transforms.py [hunk 9]
```diff
@@ -182,15 +200,16 @@ def __call__(self, sample):
...
         input_data = sample['input']
 
-        w, h = input_data.size
+        w, h = input_data[0].size
         th, tw = self.size
         fh = int(round((h - th) / 2.))
         fw = int(round((w - tw) / 2.))
 
         params = (fh, fw, w, h)
         self.propagate_params(sample, params)
```
Comment (charleygros) [inline medicaltorch/transforms.py]: Not sure: if the input contrasts do not have the same size, then these params are not the same across contrasts --> is it robust to this aspect?

File: medicaltorch/transforms.py [hunk 10]
```diff
@@ -262,7 +282,7 @@ def __init__(self, mean, std):
     def __call__(self, sample):
         input_data = sample['input']
 
-        input_data = F.normalize(input_data, self.mean, self.std)
+        input_data = [F.normalize(input_data[i], self.mean, self.std) for i in range(len(input_data))]
```
Comment (charleygros) [inline medicaltorch/transforms.py]: Not sure: I thought mean and std are lists (for each contrast)

File: medicaltorch/transforms.py [hunk 11]
```diff
@@ -301,15 +325,14 @@ class NormalizeInstance3D(MTTransform):
...
-
-        mean, std = input_data.mean(), input_data.std()
-
-        if mean != 0 or std != 0:
-            input_data_normalized = F.normalize(input_data,
-                                                [mean for _ in range(0, input_data.shape[0])],
-                                                [std for _ in range(0, input_data.shape[0])])
-
+        for i in range(len(input_data)):
+            mean, std = input_data.mean(), input_data.std()
```
Comment (charleygros) [inline medicaltorch/transforms.py:365]: Would it be better to compute the mean per channel? Not sure which way to go

File: medicaltorch/datasets.py (hunk 6 above)
Comment (AnBucquet) [reply medicaltorch/datasets.py]: Yep, thx!

File: medicaltorch/transforms.py (hunk 7 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py:56]: `ToTensor()` returns:
- a **tensor**  in case of single input
- a **list of tensors** otherwise.

All transformation (from **MedicalTorch** or **IvadoMed**) that come after the later, deal with both. As **MedicalToch** is used by other people, I've preferred to return the same type as the current master in case of a single input.

File: medicaltorch/transforms.py (hunk 8 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py:58]: input_metadata is a _list_ of _dict_ by default (c.f. [L352](https://github.com/neuropoly/medicaltorch/blob/ab/generalization_dataset/medicaltorch/datasets.py#L352)). In case of a single input, there is just one _dict_.

File: medicaltorch/transforms.py (hunk 2 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py]: I guess you're right, I'll fix it now!

File: medicaltorch/transforms.py (hunk 9 above)
Comment (AnBucquet) [reply medicaltorch/transforms.py]: Nop, you're right! What about [this](https://github.com/neuropoly/medicaltorch/blob/ab/generalization_dataset/medicaltorch/transforms.py#L215-L220)? Do we need to save the parameters for every contrast in the Ground Truth's metadata?

Comment (charleygros) [reply medicaltorch/transforms.py]: Sorry for the late reply .. ..
Looks good
"for every contrast in the Ground Truth's metadata" --> I would say: in GT_metadata --> params of gt_crop while input1_metadata --> params of input1_crop.
